### PR TITLE
Add ERC721 tests for order watcher v2

### PR DESCRIPTION
**Pull request by LogvinovLeon**
## Description

This PR:
* Adds support for ERC721 in fill scenarios
* Adds ERC721 tests to the order watcher
* Fixes ERC721 related bugs in order-watcher

## Testing instructions

Run order-watcher tests

## Types of changes

*   [ ] Bug fix (non-breaking change which fixes an issue)
*   [ ] New feature (non-breaking change which adds functionality)
*   [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

*   [ ] Change requires a change to the documentation.
*   [ ] Added tests to cover my changes.
*   [ ] Added new entries to the relevant CHANGELOG.jsons.
*   [ ] Labeled this PR with the 'WIP' label if it is a work in progress.
*   [ ] Labeled this PR with the labels corresponding to the changed package.

### CODE REVIEW

File: packages/order-watcher/test/order_watcher_test.ts [hunk 1]
```diff
@@ -534,5 +534,65 @@ describe('OrderWatcher', () => {
...
+                    await orderWatcher.addOrderAsync(signedOrder);
+                    const callback = callbackErrorReporter.reportNodeCallbackErrors(done)((orderState: OrderState) => {
+                        expect(orderState.isValid).to.be.false();
+                        const invalidOrderState = orderState as OrderStateInvalid;
+                        expect(invalidOrderState.orderHash).to.be.equal(orderHash);
+                        expect(invalidOrderState.error).to.be.equal(ExchangeContractErrs.InsufficientMakerAllowance);
+                    });
+                    orderWatcher.subscribe(callback);
+                    const isApproved = false;
+                    await contractWrappers.erc721Token.setProxyApprovalForAllAsync(
```
Comment (fabioberger) [inline packages/order-watcher/test/order_watcher_test.ts]: Let's add a test for changing the approval of a single token.

File: packages/order-watcher/tsconfig.json [hunk 2]
```diff
@@ -1,6 +1,7 @@
 {
     "extends": "../../tsconfig",
     "compilerOptions": {
+        "downlevelIteration": true,
```
Comment (fabioberger) [inline packages/order-watcher/tsconfig.json]: Let's do more research into this and if there are no large downsides, let's move to top-level `tsconfig.json`.

File: packages/fill-scenarios/src/fill_scenarios.ts [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+        expirationTimeSeconds?: BigNumber,
+    ): Promise<SignedOrder> {
+        const decodedMakerAssetData = assetProxyUtils.decodeAssetData(makerAssetData);
+        if (decodedMakerAssetData.assetProxyId === AssetProxyId.ERC20) {
+            await this._increaseERC20BalanceAndAllowanceAsync(
+                decodedMakerAssetData.tokenAddress,
+                makerAddress,
+                makerFillableAmount,
+            );
+        } else {
```
[the remark this answers is not in the document]
Comment (LogvinovLeon) [reply packages/fill-scenarios/src/fill_scenarios.ts]: The decoding function already throws. I renamed it from `decodeAssetData` to `decodeAssetDataOrThrow`

File: packages/fill-scenarios/src/fill_scenarios.ts [hunk 4]
```diff
@@ -0,0 +1,285 @@
...
+            artifacts.DummyERC721Token.compilerOutput.abi,
+            tokenAddress,
+            this._web3Wrapper.getProvider(),
+            this._web3Wrapper.getContractDefaults(),
+        );
+        const isApproved = true;
+        const txHash = await erc721Token.setApprovalForAll.sendTransactionAsync(this._erc721ProxyAddress, isApproved, {
+            from: address,
+        });
+        await this._web3Wrapper.awaitTransactionSuccessAsync(txHash);
```
[the remark this answers is not in the document]
Comment (LogvinovLeon) [reply packages/fill-scenarios/src/fill_scenarios.ts]: Yes. At the time ERC20 version was written we were not running tests against Geth. Now it's necessary.

File: packages/order-watcher/tsconfig.json (hunk 2 above)
Comment (LogvinovLeon) [reply packages/order-watcher/tsconfig.json]: https://blog.mariusschulz.com/2017/06/30/typescript-2-3-downlevel-iteration-for-es3-es5

TL;DR
The only downside is that if we run in the environment with no [Interator support](https://caniuse.com/#search=iterator) we need to include the polyfill for it.

File: packages/subproviders/tsconfig.json [hunk 5]
```diff
@@ -1,8 +1,7 @@
 {
     "extends": "../../tsconfig",
     "compilerOptions": {
-        "outDir": "lib",
-        "downlevelIteration": true
```
[the remark this answers is not in the document]
Comment (LogvinovLeon) [reply packages/subproviders/tsconfig.json]: We moved it to the global one
